### ISSUE TITLE
Automatically convert basic types to AiiDA types

### ISSUE THREAD

**Issue by greschd** [aiidateam/aiida_core#629]
When providing inputs to a ``WorkChain``, it would be nice if basic Python types (``int``, ``float``, etc.) converted to the corresponding AiIDA types automatically. A function that does this conversion is already implemented in ``data.base.to_aiida_type``. Maybe it would also be nice to make this configurable from the ``spec.input``, i.e. being able to provide a function which takes the input and converts it into an AiiDA type.

In terms of implementing this, the following would be a possible route: 
* add a keyword argument (``conversion_function``) to ``ProcessSpec.input``, which defaults to ``data.base.to_aiida_type`` (in AiiDA, not Plum)
* In ``Process.__init__``, apply the conversion functions to all inputs before anything else

What are your thoughts?

**Comment by greschd** [comment]
A little context: The reason I came up with this idea is partly because I wanted to pass a ``WorkChain`` as input. Now the best way I see to do this is to convert it into a string using ``plum.util.fullname``, and then ``plum.util.load_class`` when the input is actually used. Obviously it would be nice if the user didn't have to do this conversion himself.

So if possible it would be even nicer to have **two** conversion functions, one that turns it into a safe-able type (in that case, ``fullname``), and one that restores the original type (``load_class``).

From the user's perspective, if the ``valid_types`` are the ones **before** conversion seems logical, but I don't know if that's easily possible. I think it would require some change in plum, to enable the conversion after checking the inputs.

**Comment by broeder-j** [comment]
my 2 cents:

**Comment by greschd** [comment]
I don't think there's a good way to store generic python objects such as classes, functions, etc. The problem is that a class / function can really contain *anything*, and there's no realistic way to map that to the DB. The ``pickle`` module tries to be a sort-of-general serializer, but also fails in some instances. There's ``dill`` which is a bit better still, but it also has its limitations.

As to your second point, I think that we have always had a loose definition of *data* -- the analogous to what I'm trying to do here is having a flag ``mode``, where you can give a string that selects how a specific part of the ``WorkChain`` behaves. The input in this case is also not *data* in a strict sense. 

Now the problem with the above method (``method`` flag), is that it violates the open/closed principle: Every time I want to add a new method, I need to edit the base workchain. This quickly becomes a maintenance nightmare, since I'm writing a base workchain where parts (e.g. some DFT calculation) can be handled by different codes. As maintainer of the base workchain, I shouldn't have to care what codes the users want to be available.

I had a discussion about this proposal with @sphuber, and the conclusion was that having this automatic conversion was ok **if** it only happens when the workchain developer **explicitly** enables it. That means, the ``spec.input`` would get ``serialize_fct`` and ``deserialize_fct`` keywords, but they both default to doing nothing. 

So in conclusion, the system would remain completely backwards-compatible, and it is up to the workchain developer to decide if he want this conversion.
